File: Blind75-Practice/easy/10-backtracking/Python/sum-of-all-subsets-xor-total.py
"""
LeetCode 1863: Sum of All Subsets XOR Total
Difficulty: Easy
Pattern: Backtracking / DFS

Problem:
The XOR total of an array is the bitwise XOR of all its elements, or 0 if empty.
For every subset of nums, compute its XOR total and return the sum of those XOR totals.

Example 1:
Input: nums = [1,3]
Output: 6
Explanation: Subsets are [], [1], [3], [1,3] with XOR totals 0, 1, 3, 2; sum = 6.

Example 2:
Input: nums = [5,1,6]
Output: 28

Constraints:
- 1 <= nums.length <= 12
- 1 <= nums[i] <= 20

Time Complexity: O(2^n)
Space Complexity: O(n) recursion depth
"""


from typing import List


class Solution:
    def subsetXORSum(self, nums: List[int]) -> int:
        """DFS: at each index, branch with and without nums[i] in the current XOR."""

        def dfs(i: int, cur: int) -> int:
            if i == len(nums):
                return cur
            return dfs(i + 1, cur ^ nums[i]) + dfs(i + 1, cur)

        return dfs(0, 0)


if __name__ == "__main__":
    sol = Solution()
    assert sol.subsetXORSum([1, 3]) == 6
    assert sol.subsetXORSum([5, 1, 6]) == 28
    assert sol.subsetXORSum([3]) == 3
    print("All tests passed.")

File: Blind75-Practice/easy/10-backtracking/Java/SumOfAllSubsetsXorTotal.java
/*
 * 1863. Sum of All Subsets XOR Total
 * Difficulty: Easy
 * Pattern: Backtracking
 *
 * Problem:
 * The XOR total of an array is the bitwise XOR of all its elements, or 0 if the
 * array is empty.
 *
 * Given an integer array `nums`, consider every possible subset of `nums`.
 * Return the sum of the XOR totals of all those subsets.
 *
 * Example 1:
 * Input: nums = [1, 3]
 * Output: 6
 * Explanation:
 * Subsets are:
 * - [] -> 0
 * - [1] -> 1
 * - [3] -> 3
 * - [1, 3] -> 1 ^ 3 = 2
 * Total = 0 + 1 + 3 + 2 = 6
 *
 * Example 2:
 * Input: nums = [5, 1, 6]
 * Output: 28
 * Explanation:
 * There are 8 subsets, and the sum of all their XOR totals is 28.
 *
 * Example 3:
 * Input: nums = [3]
 * Output: 3
 * Explanation:
 * The subsets are [] and [3], so the answer is 0 + 3 = 3.
 *
 * Constraints:
 * - 1 <= nums.length <= 12
 * - 1 <= nums[i] <= 20
 *
 * Why backtracking works:
 * At each index, you have two choices: include `nums[i]` in the running XOR or skip
 * it. Recursing over those two choices explores every subset exactly once.
 *
 * Time: O(n * 2^n)
 * Space: O(n) recursion stack
 */

public class SumOfAllSubsetsXorTotal {
    public int subsetXORSum(int[] nums) {
            return dfs(nums, 0, 0);
        }
        
        private int dfs(int[] nums, int i, int curXor) {
            if (i == nums.length) {
                return curXor;
            }
            return dfs(nums, i + 1, curXor ^ nums[i])
                    + dfs(nums, i + 1, curXor);
        }
        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        // public int subsetXORSum(int[] nums) {
        //     return dfs(nums, 0, 0);
        // }
        //
        // private int dfs(int[] nums, int i, int curXor) {
        //     if (i == nums.length) {
        //         return curXor;
        //     }
        //     return dfs(nums, i + 1, curXor ^ nums[i])
        //             + dfs(nums, i + 1, curXor);
        // }
    }

    public static void main(String[] args) {
        SumOfAllSubsetsXorTotal sol = new SumOfAllSubsetsXorTotal();
        assert sol.subsetXORSum(new int[] {1, 3}) == 6;
        assert sol.subsetXORSum(new int[] {5, 1, 6}) == 28;
        assert sol.subsetXORSum(new int[] {3}) == 3;
        System.out.println("All tests passed.");
    }
}
